RDRP-1002: freezing form sent out bug

JenCheshire opened the pull request:
Initial bug: where status is form sent out in frozen data but is a clear response in the update, the update was coming in as an addition but not dropping the original row.  This has now been fixed by dropping the references that have been replaced by an addition (for specific form types).

Whilst investigating this, another bug was found re: check needed and updated/additional instances.  These are now all dealt with in amendments - there is a check for refs that appear in both - new instances are moved to the amendments file so they can all be considered together.  Redundant validation has been removed.

Updated addition/amendment files have examples of these cases :)

## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #


#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)


#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

JenCheshire committed:
RDR-1002: update to amends

JenCheshire committed:
RDRP-1002: removed redundant validation

JenCheshire committed:
RDRP-1002: precommit changes

JenCheshire committed:
RDRP-1002: add specific status to drop

JenCheshire committed:
RDRP-1002: added test for split cases function

zorge69 commented on src/freezing/freezing_compare.py:
```diff
@@ -200,3 +208,52 @@ def output_freezing_files(
+        amendments_df (pd.DataFrame): The records that have changed.
+        FreezingLogger (logging.Logger): The logger to log to.
+
+    Returns:
```
Actually, it returns additions_df, amendments_df. Please update the doc string.

zorge69 commented on src/freezing/freezing_compare.py:
```diff
@@ -200,3 +208,52 @@ def output_freezing_files(
         return True
+
+
+def check_for_split_cases(additions_df, amendments_df, FreezingLogger):
```
Please add output hit type (tuple of dataframes)

zorge69 commented on tests/test_freezing/test_freezing_compare.py:
```diff
@@ -5,6 +5,7 @@
 import logging
```
does it have to be pandas.testing
 without the underscore?

zorge69 commented on tests/test_freezing/test_freezing_compare.py:
```diff
@@ -167,3 +168,80 @@ def test_get_additions(self):
+        # Check the output
+        assert_frame_equal(
+            expected_additions_df, result_additions_df.reset_index(drop=True)
+        ), "check_for_split_cases() additions output is not as expected."
```
This may be right, but I don't get it. Why is there a string after the comma?
), "get_additions() output is not as expected."

JenCheshire replied:
It's a custom error message for if the assert fails :)

zorge69 commented on tests/test_freezing/test_freezing_compare.py:
```diff
@@ -167,3 +168,80 @@ def test_get_additions(self):
+
+        assert_frame_equal(
+            expected_amendments_df, result_amendments_df.reset_index(drop=True)
+        ), "check_for_split_cases() amendments output is not as expected."
```
same as above

zorge69 reviewed (changes requested):
The unit test passes. The pipeline runs end to end. The function check_for_split_cases is doing the right thing (name is not ideal, as it not only checks, but transfers refs).

Please answer my questions and I will be happy to approve,

JenCheshire committed:
RDRP-1002: changes based on feedback

JenCheshire commented:
Thanks @zorge69 - I have changed the name of the function.  Just relooked at those statements, and you're right, they are not doing anything so I have removed

JenCheshire committed:
RDRP-1002: remove statements
